onnxruntime/core/providers/webnn/builders/helper.cc

```diff
@@ -99,6 +99,70 @@ bool IsTensorShapeSupported(const NodeArg& node_arg, const std::string& parent_n
   return true;
 }
 
+// Check if all input tensor ranks of the given node are supported by WebNN.
+bool IsInputRankSupportedByOp(const Node& node, const emscripten::val& wnn_limits, const logging::Logger& logger) {
+  const std::string_view op_type = node.OpType();
+
+  auto it = op_inputs_map.find(op_type);
+  if (it == op_inputs_map.end()) {
+    LOGS(logger, VERBOSE) << "Operator type: [" << op_type << "] is not found in the op inputs map.";
+    return false;
+  }
+
+  for (const auto& input : it->second.inputs) {
+    const std::string_view webnn_op_type = it->second.opType;
+
+    if (static_cast<size_t>(input.index) >= node.InputDefs().size() || node.InputDefs()[input.index] == nullptr) {
+      LOGS(logger, VERBOSE) << "Operator type: [" << op_type
+                            << "], input index: [" << input.index
+                            << "], corresponding WebNN op type: " << webnn_op_type
+                            << ", WebNN input name " << input.name
+                            << "] is not found.";
+      return false;
+    }
+
+    std::vector<int64_t> input_shape;
+    if (!GetShape(*node.InputDefs()[input.index], input_shape, logger)) {
+      return false;
+    }
+
+    if (wnn_limits[std::string(webnn_op_type)].isUndefined() ||
+        wnn_limits[std::string(webnn_op_type)][std::string(input.name)].isUndefined()) {
+      LOGS(logger, VERBOSE) << "Operator type: [" << op_type
+                            << "], input index: [" << input.index
+                            << "], corresponding WebNN op type: " << webnn_op_type
+                            << ", WebNN input name " << input.name
+                            << " is not defined in wnn_limits.";
+      return false;
+    }
+
+    const auto& input_limits = wnn_limits[std::string(webnn_op_type)][std::string(input.name)];
```

Comment:
If we really need to wrap the `string_view` with a `string` just to perform a lookup, then better to do it exactly once outside the loop rather than multiple times (which could allocate memory). Ideally I'd like `wnn_limits` to take advantage of C++14's [heterogeneous lookups](https://www.cppstories.com/2021/heterogeneous-access-cpp20/), but the object comes from emscripten, and it seems to not support that sadly.